src/main/java/io/synadia/flink/source/NatsJetStreamSourceReader.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.flink.core.io.InputStatus;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+public class NatsJetStreamSourceReader<OutputT> implements SourceReader<OutputT, NatsSubjectSplit> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(NatsJetStreamSourceReader.class);
+
+    private final String id;
+    private final ConnectionFactory connectionFactory;
+    private final DeserializationSchema<OutputT> payloadDeserializer;
+    private final SourceReaderContext readerContext;
+    private final List<NatsSubjectSplit> subbedSplits;
+    private final FutureCompletingBlockingQueue<Message> messages;
+    private Connection connection;
+    private Subscription subscription;
+    private NatsConsumeOptions config;
+    private JetStream js;
+    private String subject;
+    private final Boundedness mode;
+    public NatsJetStreamSourceReader(String sourceId,
+                                     ConnectionFactory connectionFactory,
+                                     NatsConsumeOptions natsConsumeOptions,
+                                     DeserializationSchema<OutputT> payloadDeserializer,
+                                     SourceReaderContext readerContext,
+                                     String subject,
+                                     Boundedness mode) {
+        id = sourceId + "-" + Utils.generatePrefixedId(sourceId);
+        this.connectionFactory = connectionFactory;
+        this.payloadDeserializer = payloadDeserializer;
+        this.readerContext = checkNotNull(readerContext);
+        subbedSplits = new ArrayList<>();
+        messages = new FutureCompletingBlockingQueue<>();
+        this.config= natsConsumeOptions;
+        this.subject = subject;
+        this.mode = mode;
+    }
+
+
+    @Override
+    public void start() {
+        LOG.debug("{} | start", id);
+        try {
+            connection = connectionFactory.connect();
+            js = connection.jetStream();
+            PullSubscribeOptions pullOptions = PullSubscribeOptions.builder()
+                    .stream(config.getStreamName()).bind(true).durable(config.getConsumerName())
+                    .build();
+            subscription = js.subscribe(subject, pullOptions);
+        }
+        catch (IOException e) {
+            throw new FlinkRuntimeException(e);
+        } catch (JetStreamApiException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput<OutputT> output) throws Exception {
+        List<Message> messages =
+                ((JetStreamSubscription) subscription).fetch(config.getBatchSize(), Duration.ofSeconds(5));
```

Comment:
`maxWait` here is hard-coded to 5s. I think it could be configurable?

Reply:
// TODO

Reply:
Agreed, @somratdutta please make it configurable.

Reply:
sure will do it